_posts/2015-12-13-testing-multithreaded-code-in-java.md

```diff
@@ -0,0 +1,154 @@
+---
+layout: post
+name: Testing-multithreaded-code-in-java
+title: 'Testing multithreaded code in Java'
+date: 2015-12-13 00:20:00 +00:00
+author: Felipe Fernández
+image:
+    src: /assets/img/custom/blog/picot.png
+tags:
+- testing
+- concurrency
+- java
+
+---
+
+Testing multithreaded code is a tough challenge. The first advice that you get when try to test concurrency is to isolate your concurrent concerns in the code as much as possible. This a general design advice but in this case it's even more important. When you have properly unit tested the logic that is wrapped by the concurrent construct, you can start thinking about how to test that bit. Otherwise you can spend so many time trying to figure out what's the concurrent problem, being in the end a simple flawed business logic.
+
+Once you got that covered, you can think in your strategy to test concurrent systems. [Goos](http://www.growing-object-oriented-software.com/) covers how you can do it. [Here](https://github.com/npryce/goos-code-examples/tree/master/testing-multithreaded-code/src/book/example/threading/races) you can find the code that I'm going to explain:
+
+First, let's have a look into the SUT:
+
+```
+    public class AtomicBigCounter {
+        private BigInteger count = BigInteger.ZERO;
+
+    	public BigInteger count() {
+        	return count;
+    	}
+
+    	public void inc() {
+        	count = count.add(BigInteger.ONE);
+    	}
+	}
+```  
+
+As you can see, this class is not thread safe, as it's exposing some state through inc method, being that state not thread safe (you could use AtomicInteger instead of BigInteger, to make it thread safe). To test that class we'll include a non-concurrent and a concurrent test.
```

Comment:
some state through the ```inc()``` method. The state is not thread safe (you could use AtomicInteger instead of BigInteger to fix that).